did-doc-diff/src/index.ts

```diff
@@ -0,0 +1,217 @@
+      "key": "did:key:zDnaesnjCLbxSB2EiJgTZUYCBqHTkuDJDdLbSmaAn2bpqXtuY",
+      "sig": "z4Y3LwMCQcmgFvyGfMmSCa5Qr7vCr7tUUDWnRBQqzZsDpSCKP3ZB1WnQTUVBivruavBsCprhjJ5p4WBR1RTyeYqUd"
+    }
+
+  This makes
+    {
+    id: 'did:aic:zwulnkrxrjkobowd',
+    a: 1,
+    c: 3,
+    d: 4,
+    e: { ea: 'each' },
+    }
+*/
+
+export async function tick_to_diffs(tick: {
+  sig: string
+  key: string
+  [index: string]: any
+}) {
+  if (await validate_sig(tick)) {
+    return tick.diffs
+  } else {
+    return undefined
+  }
+}
+
+export async function tick_from_diffs(
+  diffs: { [index: string]: any },
+  key: crypto.EcdsaKeypair,
+) {
+  // This function dose not validate the diffs remember to do that first
+  const tids = Object.keys(diffs).sort()
+  if (tids.length < 1) {
+    return undefined // ToDo: handel the case where aic has no diffs
+  }
+
+  const tick = {
+    tid: TID.next().formatted(),
+    diffs: diffs,
+    id:
+      'did:aic:' +
+      (await pid(uint8arrays.fromString(JSON.stringify(diffs[tids[0]])))),
+    key: key.did(),
+    sig: '',
+  }
+  const signed_tick = sign(tick, key)
+  return signed_tick
+}
+
+function authorise_key(doc: { [index: string]: any }, didkey: string) {
+  const valid = Boolean(
+    (doc['adx/account_keys'] && doc['adx/account_keys'].includes(didkey)) ||
+      (doc['adx/recovery_keys'] && doc['adx/recovery_keys'].includes(didkey)),
+  )
+  console.log('authorise_key:', valid)
+  return true
+}
+
+export async function diffs_to_did_doc(diffs: { [index: string]: any }) {
+  if (typeof diffs !== 'object') {
+    return undefined
+  }
+  const tids = Object.keys(diffs).sort()
+  if (tids.length < 1) {
+    return undefined
+  }
+  let last = tids[0]
+  let doc = diffs[last]
+
+  // add the id feild
+  doc['id'] =
+    'did:aic:' +
+    (await pid(uint8arrays.fromString(JSON.stringify(diffs[last]))))
+
+  for (const tid of tids.slice(1)) {
+    const diff = diffs[tid]
+    if (!(await validate_sig(diff))) {
+      console.log('Warning ignoreing: bad sig')
+      continue // not valid ignore it
+    }
+    if (!authorise_key(doc, diff.key)) {
+      console.log('Warning ignoreing: bad key')
+      continue // not valid ignore it
+    }
+
+    const { prev, patches } = diff
+    if (prev !== last) {
+      console.log('Warning ignoreing: out of order')
+      continue
+    }
+    doc = patch(doc, patches)
+    last = tid
+  }
+  return doc
+}
+
+const S32_CHAR = '234567abcdefghijklmnopqrstuvwxyz'
+const B32_CHAR = 'abcdefghijklmnopqrstuvwxyz234567'
+
+function s32encode(data: Uint8Array): string {
+  // this is gross so many alocatons and funtion calls :(
+  const base32 = uint8arrays.toString(data, 'base32')
+  const sort_order_invariant_base32 = base32.split('').map((c) => {
+    return S32_CHAR[B32_CHAR.indexOf(c)]
+  })
+  return sort_order_invariant_base32.join('')
+}
+
+async function pid(data: Uint8Array) {
+  // return the pid120 a 24 char pid
+  const hash = s32encode((await sha256.digest(data)).digest)
```

Comment:
this works. but we also have a `sha256` in `@adxp/crypto` that would prevent the extra dependency

Reply:
thanks, used  it